rosbag2_cpp/include/rosbag2_cpp/message_definitions/local_message_definition_source.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright 2022, Foxglove Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__MESSAGE_DEFINITIONS__LOCAL_MESSAGE_DEFINITION_SOURCE_HPP_
+#define ROSBAG2_CPP__MESSAGE_DEFINITIONS__LOCAL_MESSAGE_DEFINITION_SOURCE_HPP_
+
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "rosbag2_cpp/visibility_control.hpp"
+#include "rosbag2_storage/message_definition.hpp"
+
+namespace rosbag2_cpp
+{
+
+class DefinitionNotFoundError : public std::exception
+{
+private:
+  std::string name_;
+
+public:
+  explicit DefinitionNotFoundError(std::string name)
+  : name_(std::move(name))
+  {
+  }
+
+  const char * what() const noexcept override
+  {
+    return name_.c_str();
+  }
+};
+
+class LocalMessageDefinitionSource final
+{
+public:
+  /**
+   * Concatenate the message definition with its dependencies into a self-contained schema.
+   * The format is different for MSG and IDL definitions, and is described fully in
+   * docs/message_definition_encoding.md
+   * Throws DefinitionNotFoundError if one or more definition files are missing for the given
+   * package resource name.
+   */
+  rosbag2_storage::MessageDefinition get_full_text(const std::string & type_name);
+
+  enum struct Format
+  {
+    MSG,
```

Comment:
@james-rms Could you please add `UNKNOWN` as first item in the enum?
```suggestion
   UNKNOWN = 0,
    MSG,
```